File: 530-minimum-absolute-difference-in-bst/530.py
class Solution:
    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)
        if self.prev is not None:
            self.curr_min = min(self.curr_min, root.val - self.prev)
        self.prev = root.val
        self.dfs(root.right)

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.curr_min = math.inf
        self.prev = None
        self.dfs(root)

        return self.curr_min

File: 530-minimum-absolute-difference-in-bst/530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* prev;
    TreeNode* curr;
    int diff;
    void dfs(TreeNode* root) {
        if (!root)
            return;
        dfs(root->left);
        if (prev) {
            diff = min(diff, root->val - prev->val);
            prev = root;
        } else {
            prev = root;
        }
        dfs(root->right);
    }
public:
    int getMinimumDifference(TreeNode* root) {
        diff = INT_MAX;
        dfs(root);
        return diff;
    }
};
